store: simplify imagestore implementation

giuseppe opened the pull request:
simplify the imagestore implementation and use it as a first-class store.  It reduces the amount of code to deal with an image store in the driver (only overlay supports it at the moment).

Closes: https://github.com/containers/storage/issues/1779

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

rhatdan commented:
Can you run a test against podman as well?

giuseppe commented:
this is ready for review.  Both Podman and Buildah pass

Podman test PR: https://github.com/containers/podman/pull/21070
Buildah test PR: https://github.com/containers/buildah/pull/5236

rhatdan commented:
LGTM
@saschagrunert @mtrmac @nalind @vrothberg PTAL

mtrmac commented on drivers/vfs/driver.go:
```diff
@@ -204,18 +203,33 @@ func (d *Driver) create(id, parent string, opts *graphdriver.CreateOpts, ro bool
+		homedir = d.homes[0]
+	}
+	newpath := filepath.Join(homedir, d.String(), "dir", filepath.Base(id))
+	if _, err := os.Stat(newpath); err != nil {
```
⚠️ How does this handle _any_ locking? The generic code maintains RW/RO locks on the individual images / layers. This… does nothing?

giuseppe replied:
this is still outstanding.

I don't see how this is different than the previous version, what locking do we need here?

mtrmac reviewed (commented):
There are parts of this, like removing all the `d.imageStore` conditions throughout overlay, which I very much like.

On the general principle, I suppose that’s pre-existing, but I just can’t see how this works. All of the read-only stores need to be RO locked, don’t they? And locks are handled in the generic code, drivers can’t do that for themselves. So _any_ kind of iteration over stores in `dir2` seems completely broken to me. 

So… what’s really going on? Is `dir2` only to handle cross-store parent layers during an overlay mount, or something? If that were so, the driver code should not be touching the other stores during any other operation; and still that cross-store parent situation would need to be somehow explicitly tracked and locked.

I’m afraid I can’t do a full review now. Maybe there is some way to split this into much smaller PRs?

mtrmac commented on layers.go:
```diff
@@ -1056,10 +1063,17 @@ func (s *store) newLayerStore(rundir string, layerdir string, driver drivers.Dri
-	defer rlstore.stopWriting()
 	lw, err := rlstore.lockfile.GetLastWrite()
 	if err != nil {
 		return nil, err
```
(In general I don’t feel strongly about having a single RW/RO constructor, or two separate RW/RO constructors. But if this change is not _necessary_ for the primary goal of the PR, it should be in a separate PR or at least a separate commit.)

giuseppe commented on store.go:
```diff
@@ -1118,9 +1116,18 @@ func (s *store) getROLayerStoresLocked() ([]roLayerStore, error) {
+
+		ro := store != s.imageStoreDir && store != s.graphRoot
+
+		rls, err := s.newLayerStore(rlpath, glpath, s.graphDriver, false, ro)
```
this is needed because we have this weird specification that a layer must be deleted from the current store if it is found there.  We even have a test for it :/

https://github.com/containers/storage/blob/main/tests/split-store.bats#L96-L101

I am fine if we decide to break this specification, but let's do it separately after we verify with CRI-O folks that it is not necessary.

mtrmac replied:
Fine… but I can’t see how that works. `layersToRemove` is only applied to one `rlstore`. That uses `getLayerStoreLocked`, and AFAICS never gets to write to this special member of `roLayerStoresUseGetters`.

I’m probably being stupid.

giuseppe commented on drivers/vfs/driver.go:
```diff
@@ -31,8 +31,9 @@ func init() {
 	d := &Driver{
 		name:       "vfs",
-		homes:      []string{home},
+		homes:      []string{filepath.Dir(home)},
```
this is the same assumption done here: https://github.com/containers/storage/blob/main/drivers/vfs/driver.go#L209-L211

the vfs driver is already assuming the vfs home directory is `$GRAPH_ROOT/"vfs"`

mtrmac replied:
Yes, from a quick skim it seems to me that the logic of the `imagestore` options would be better moved to the generic code.

that would probably be fairly invasive… a separate PR?

mtrmac commented on drivers/vfs/driver.go:
```diff
@@ -47,7 +48,7 @@ func Init(home string, options graphdriver.Options) (graphdriver.Driver, error)
 		switch key {
 		case "vfs.imagestore", ".imagestore":
-			d.homes = append(d.homes, strings.Split(val, ",")...)
+			d.additionalHomes = strings.Split(val, ",")
```
Can the `imagestore` option be provided twice? This would discard the first value. (“That’s invalid and should be rejected” seems a perfectly fine answer to me. “That’s undocumented but we use all of them” constraints us long-term but works fine. Silently discarding data seems not ideal.)

giuseppe replied:
good observation, I've restored the previous code

giuseppe commented on store.go:
```diff
@@ -701,6 +701,8 @@ type store struct {
 	additionalUIDs *idSet // Set by getAvailableIDs()
 	additionalGIDs *idSet // Set by getAvailableIDs()
+
+	additionalImageStores []string // Set by load().  List of additional stores including the image store dir if configured.
```
done

giuseppe commented on drivers/overlay/overlay.go:
```diff
@@ -1223,26 +1176,39 @@ func (d *Driver) getLower(parent string) (string, error) {
+
+	newpath := path.Join(homedir, id)
+
 	if _, err := os.Stat(newpath); err != nil {
```
this is still outstanding.

I don't see how this is different than the previous version, what locking do we need here?

mtrmac replied:
Yes, see the top-level comment https://github.com/containers/storage/pull/1784#pullrequestreview-1809634941 .

One immediate thing that springs to mind is deleting the parent layer while the child is being created/mounted/unmounted. I don’t know that we can fully defend against those kinds of things (we don’t hold layers locked for the whole time a child is mounted, and mount reference counts are, I think, basically per-machine and not visible over network shares).

Or maybe a layer with the same ID (pulled layers have deterministic IDs) being simultaneously created in both stores, making the path returned by `d.dir[2]()` change over time — e.g. for `getStagingDir` as an example.

There might well be more, or maybe these are handled by some mechanism I don’t know about.

---

Assuming the concern is real:
- I’d like to see the scope of the problem restricted to the unavoidable minimum (whatever that is); in all situations where the generic code is iterating over `getROLayerStoresLocked` to find the right layer, and handling locking, the driver should not be iterating over the secondary layer stores again / redundantly without locking. Structure the code, or document it, in a way that is likely to be maintained correctly over time.
- In the situations where the problem remains outstanding afterwards and is not fixable short-term, we should loudly document the problem, the situations where it arises, and the constraints, for future maintainers (compare `LOCKING BUG` in `layers.go`)
- Alternatively, if this is never going to be fixed, and users need to use the feature appropriately (“the image store must never be modified while it is used with outer stores” ?), clearly document that in end-user documentation, and maybe point at it in the code to acknowledge and justify the unsynchronized accesses.
- (As another option, a few years ago overlay used to do its own per-layer locking. I have removed that because I thought it’s redundant; actually external users can get a pointer to the driver object, so it is not _100%_ redundant, and doing that locking here might also be an option. I would prefer not to add that cost to every operation.)

I could see an argument that this is out of scope of this PR … I’m thinking that with this touching `dir2` and `dir`, while the feature is still _comparatively_ new , now is the best chance for fixing this we’ll ever have. Admittedly the timing is not super convenient now, but it will probably never be ideal.

giuseppe replied:
the assumption with additional image stores is that images/layers are not removed.  There are cases where we don't grab any lock, e.g. a read-only file system, so that is already the current state.

When we use a read-only layer, the layer could disappear also while a container is running, and there is nothing that prevents the removal now:

```
# mkdir -p /tmp/store/{rw,ro,empty}
# podman --root /tmp/store/rw pull busybox
# mount --bind --read-only /tmp/store/rw /tmp/store/ro
# podman --root /tmp/store/empty --storage-opt overlay.imagestore=/tmp/store/ro run -d busybox top
5d4baa2643c77deca49ff9b1a0a10df5c2965626b725b8a2225db54706e69980
# podman --root /tmp/store/rw rmi busybox
Untagged: docker.io/library/busybox:latest
Deleted: 9211bbaa0dbd68fed073065eb9f0a6ed00a75090a9235eca2554c62d1e75c58f
```

mtrmac replied:
Yeah, the silent deletion of dependent parents seems hard to fix. It’s also not really documented that the users must prevent that, AFAICT.

What about concurrent creations, and other operations?

giuseppe replied:
I've not thought about them, but are they very different than delete though?  We would end up with the same layer in multiple stores and use it from the first one where we find it.

mtrmac replied:
We could see e.g. a partial pull of a layer being started on our store, triggering the creation of a staging directory; before it finishes, the other image store creates the same layer; and now the code trying too commit the partially-pulled layer sees the new layer, determines a different staging directory path, and rejects the commit attempt. (I’m not saying that this is the _only_ problem.)

mtrmac commented:
If nothing else, https://github.com/containers/storage/pull/1784#discussion_r1447591983 is still outstanding. I didn’t look further.

mtrmac commented:
My opinion is that yes, I think we need to hold (or outright reject) a PR that adds a code with no known purpose.

If there is a bug to fix, and that code is not necessary to fix the bug, I’d much prefer a PR that _only_ fixes that bug. Or refactors in justifiable ways and fixes that bug.

(And if this PR was originally motivated by a bug report, nothing to this point has said so. And that rather changes both the priority and approach to the PR review…)

mtrmac commented:
(Note that the review comment I pointed at is not about locking insider the overlay driver; it’s about the new unused “second writeable layer store”.)

giuseppe commented:
> (Note that the review comment I pointed at is not about locking insider the overlay driver; it’s about the new unused “second writeable layer store”.)

we need to use the same kind of lock to access a store otherwise 
 getLockfile (https://github.com/containers/storage/blob/main/pkg/lockfile/lockfile.go#L288-L313)  fails.  Since we are accessing the store in write mode sometimes, we need to ensure we always grab the same lock kind, even if we use it as an additional store.

mtrmac commented:
> Since we are accessing the store in write mode sometimes,

Where? AFAICS this doesn’t add a single new write access to a layer store to `store.go` (nor any new `rwLayerStore.startWriting` call); it _just_ changes the constructor.

Would we somehow have two different stores, with two different `imageStoreDir`/`graphRoot` configurations in a single process?

giuseppe commented:
> Would we somehow have two different stores, with two different `imageStoreDir`/`graphRoot` configurations in a single process?

I've dropped the part where the imagestore/graphroot was used in `getROLayerStoresLocked()`, so this part should be less confusing now.

I've added some patches to handle locking for the imagestore, so both the imagestore and the graphroot are locked when an imagestore is used through `bothLayerStoreKind()`, but I need to make another pass at these patches to see what more I can simplify/clean up.

giuseppe commented:
@rhatdan @mtrmac can we please move this forward?

rhatdan commented:
@nalind PTAL

giuseppe commented:
> Is this related to the Zstd work?

no, this is not related to zstd

rhatdan commented:
@giuseppe @mtrmac is this something we should get in before 5.0?

giuseppe commented:
I don't think it is too important for Podman 5.0, but it solves an issue that was reported some time ago: https://github.com/containers/storage/issues/1779

giuseppe commented:
@kolyshkin @nalind PTAL

rhatdan commented:
@saschagrunert PTAL

giuseppe commented:
@haircommander @saschagrunert this patch affects the "split store" functionality in CRI-O

haircommander commented:
cc @kannon92

giuseppe commented:
rebased

rhatdan commented:
/lgtm
